prov/util/src/util_av.c

```diff
@@ -1056,92 +1080,114 @@ static int ofi_cmap_add_peer(struct util_cmap *cmap, struct util_cmap_handle *ha
 	if (dlist_find_first_match(&cmap->peer_list, ofi_cmap_match_peer, addr)) {
 		FI_WARN(cmap->av->prov, FI_LOG_EP_CTRL,
 				"Peer already present\n");
-		goto out;
+		ret = -FI_EINVAL;
+		goto unlock;
 	}
 
 	// TODO Use util_buf_pool
 	peer = calloc(1, sizeof(*peer) + addrlen);
 	if (!peer) {
 		ret = -FI_ENOMEM;
-		goto out;
+		goto unlock;
 	}
 
 	ofi_cmap_init_handle(handle, cmap, state, FI_ADDR_UNSPEC, peer);
 	peer->handle = handle;
 	peer->addrlen = addrlen;
 	memcpy(peer->addr, addr, addrlen);
 	dlist_insert_tail(&peer->entry, &cmap->peer_list);
-out:
+unlock:
 	fastlock_release(&cmap->lock);
 	return ret;
 }
 
-/*
- * Caller must hold cmap->lock. Either fi_addr or
- * addr and addrlen args should be present.
- */
+/* Either fi_addr or addr and addrlen args must be given. */
 int ofi_cmap_add_handle(struct util_cmap *cmap, struct util_cmap_handle *handle,
 		enum util_cmap_state state, fi_addr_t fi_addr, void *addr,
 		size_t addrlen)
 {
-	int index;
+	int index, ret = 0;
 	if (fi_addr == FI_ADDR_UNSPEC) {
 		index = ip_av_get_index(cmap->av, addr);
 		if (index < 0)
 			return ofi_cmap_add_peer(cmap, handle, state, addr, addrlen);
 		fi_addr = index;
 	}
 
-	if (cmap->handles[fi_addr]) {
-		FI_WARN(cmap->av->prov, FI_LOG_EP_CTRL,
-				"Handle already present\n");
-	} else {
-		ofi_cmap_init_handle(handle, cmap, state, fi_addr, NULL);
-		cmap->handles[fi_addr] = handle;
+	if (cmap->handles_av[fi_addr]) {
+		FI_TRACE(cmap->av->prov, FI_LOG_EP_CTRL, "Handle already present\n");
```

Comment:
I added it as a TRACE because when two threads try to add a handle (initiate connection to a remote EP), one of them would find the handle already added by another thread. So it wouldn't be a critical warning, right?